backend/src/main/java/gov/cdc/usds/simplereport/logging/PatientExperienceLoggingInterceptor.java

```diff
@@ -0,0 +1,65 @@
+package gov.cdc.usds.simplereport.logging;
+
+import gov.cdc.usds.simplereport.api.pxp.CurrentPatientContextHolder;
+import gov.cdc.usds.simplereport.db.model.PatientLink;
+import gov.cdc.usds.simplereport.service.AuditService;
+import java.util.UUID;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.slf4j.MDC;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.stereotype.Component;
+import org.springframework.web.servlet.HandlerInterceptor;
+import org.springframework.web.servlet.ModelAndView;
+
+/**
+ * HandlerInterceptor to set a request ID for patient experience REST handlers, as well as logging
+ * audit information for patient requests.
+ */
+@Component
+@ConditionalOnWebApplication
+public class PatientExperienceLoggingInterceptor implements HandlerInterceptor {
+
+  private static final Logger LOG = LoggerFactory.getLogger(PatientExperienceLoggingInterceptor.class);
+
+  @Autowired private CurrentPatientContextHolder _context;
+  @Autowired private AuditService _auditService;
+
+  @Override
+  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
+      throws Exception {
+    LOG.debug(
+        "Pre-handling a method={} request for uri={} using handler={}",
+        request.getMethod(),
+        request.getRequestURI(),
+        handler);
+    String requestId = UUID.randomUUID().toString();
+    MDC.put(GraphQLLoggingHelpers.GRAPHQL_QUERY_MDC_KEY, requestId);
+    return true;
+  }
+
+  @Override
+  public void postHandle(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      Object handler,
+      ModelAndView modelAndView)
+      throws Exception {
+    PatientLink patientLink = _context.getPatientLink();
+    LOG.debug("Closing out request. Patient link is {}", patientLink);
+    if (patientLink != null) { // this is noooot acceptable
+      String requestId = MDC.get(GraphQLLoggingHelpers.GRAPHQL_QUERY_MDC_KEY);
+      _auditService.logRestEvent(requestId, request, _context.getOrganization(), patientLink);
+    }
```

Comment:
Should this maybe log a stern warning and capture an audit event with a missing patient link if `patientLink` is `null`?

Reply:
That seems like an important question. My basic thinking is

1) this is to capture actions by patients
2) absent a PatientLink we cannot identify the patient
3) AU-5 says we bail if we can't record things
4) now therefore, the absence of a validated PatientLink should be a fatal error and we should kick back some 4xx error and call it a day.

This is slightly complicated by 
1. the fact that we are sitting _above_ the authentication layer, so something can come back to us with no patient link but a 403 error already, which is something we neither want to change or log; and
2. the recent introduction of legitimately anonymous endpoints (the registration form).

My solution for the second one is to explicitly identify anonymous-OK endpoints, either by explicit whitelist or by a marker interface or annotation (the last one being my general preference). With that done, anybody who has no PatientLink gets the greater of 400 and whatever they already had, and I think it more or less works out.

Reply:
I think for anonymous endpoints and failed authentication attempts, we can't identify the requestor. We only need to bail if writing an audit event fails, not if there's some information which can't be known for a the action being recorded.

Reply:
Oh, I suppose we do probably want to record the an audit event for the anonymous user, at that. Too bad I made api_user_id a non-null field! &lt;sight&gt;

Logging authentication failures introduces an asymmetry, though: we don't log 403 or 401 responses on the graphql endpoint, because we never see them. But then, Okta probably does log auth failures on their side. So should we be logging those as well, here?